shared/common-adapters/header-hoc.native.js

```diff
@@ -4,54 +4,127 @@ import Text from './text'
 import {StyleSheet} from 'react-native'
 import BackButton from './back-button'
 import Box from './box'
+import FloatingMenu from './floating-menu'
+import Icon from './icon'
 import * as Styles from '../styles'
-import type {Props} from './header-hoc.types'
+import type {Action, Props} from './header-hoc.types'
 
-export const HeaderHocHeader = ({
-  headerStyle,
-  customComponent,
-  hideBackLabel,
-  title,
-  onCancel,
-  customCancelText,
-  onBack,
-  onRightAction,
-  rightActionLabel,
-  theme = 'light',
-}: Props) => (
-  <Box style={Styles.collapseStyles([styles.header, theme === 'light' ? styles.headerLight : styles.headerDark, headerStyle])}>
-    {customComponent}
-    {!!title && (
-      <Box style={styles.titleContainer}>
-        <Text type="BodySemibold" style={styles.title}>{title}</Text>
-      </Box>
-    )}
-    {onCancel && (
-      <Text type="BodyBigLink" style={styles.button} onClick={onCancel}>
-        {customCancelText || 'Cancel'}
-      </Text>
-    )}
-    {onBack && (
-      <BackButton
-        hideBackLabel={hideBackLabel}
-        iconColor={theme === 'light' ? Styles.globalColors.black_40 : Styles.globalColors.white}
-        style={styles.button}
-        onClick={onBack}
-      />
-    )}
-    {!!rightActionLabel && (
-      <Box style={styles.rightAction}>
-        <Text
-          type="BodyBigLink"
-          style={Styles.collapseStyles([styles.button, {opacity: onRightAction ? 1 : 0.3}])}
-          onClick={onRightAction}
+const MAX_RIGHT_ACTIONS = 3
+type State = {|
+  floatingMenuVisible: boolean,
+|}
+export class HeaderHocHeader extends React.Component<Props, State> {
+  state = {
+    floatingMenuVisible: false,
+  }
+  _onHidden = () => this.setState({floatingMenuVisible: false})
+  _showFloatingMenu = () => this.setState({floatingMenuVisible: true})
+  render() {
+    // TODO: remove these after updates are fully integrated
+    const onLeftAction = this.props.onLeftAction || this.props.onBack || this.props.onCancel
+    const leftAction = this.props.leftAction || this.props.onCancel ? 'cancel' : this.props.onBack ? 'back' : null
+    const rightActions = this.props.rightActions || (this.props.onRightAction && this.props.rightActionLabel) ? [{
+      label: this.props.rightActionLabel,
+      onPress: this.props.onRightAction,
+    }] : null
+
+    return (
+      <Box
+        style={Styles.collapseStyles([
+          styles.header,
+          this.props.borderless && styles.borderless,
+          this.props.headerStyle,
+        ])}
+      >
+        {this.props.customComponent}
+        <Box style={styles.leftAction}>
```

Comment:
This render function is pretty big. I think it would do us good to split this into sub components. It would also make it easier to parse and understand. How about a sub component for each nitty gritty thing you have to do, and this render fn could match more what the user sees. Something like:

```
return (
  <Box ...>
    {this.props.customComponent}
    <LeftActions ... />
    <Title ... />
    <RightActions ... />
    <OverflowRightActions ... />
</Box>
)
```

Then each subcomponent can be clearer as well with named variables instead of chaining expressions together.

Reply:
Works for me. Thanks for the feedback!